XMonad/Actions/Menu/Core.hs

```diff
@@ -0,0 +1,116 @@
+{-|
+
+Module      : XMonad.Actions.Menu.Core
+Description : The core functionality that runs menus.
+Copyright   : (C) David Janssen, 2016
+License     : BSD3
+Maintainer  : David Janssen <janssen.dhj@gmail.com>
+Stability   : unstable
+Portability : unportable
+
+This module contains the code that is responsible for listening to the keyboard
+and running mappings between key-strokes and actions.
+
+I am thinking just now (at the time of making this 'publication-ready') that
+'doing-things-in-the-context-of-the-keyboard', which is defined by withKB and
+the various KeyStroke based actions is probably a Monad (since they're supposed
+to be 'computational-contexts', if I understand correctly.) I could probably
+refactor this into something nicer with that.
```

Comment:
I was trying to implement this today, and all of the stuff you suggested was working out well, until I tried to use `cleanMask` inside `waitKey`.

```

-- | waitKey returns the next KeyStroke that occurs
waitKey :: KB KeyStroke
waitKey = do
  XConf { display=dpy } <- ask
  (sym, e) <- io $
    -- X-stuff to grab the next (Modifiers, KeyPress) from the system
    allocaXEvent $ \e -> do
        maskEvent dpy keyPressMask e
        ev <- getEvent e
        (ks, _) <- lookupString $ asKeyEvent e
        return (fromMaybe xK_VoidSymbol ks, ev)
  -- Remove unused modifiers
  mods <- cleanMask . ev_state $ e
  return (mods, sym)
```

It now complains that it's getting an `X KeyStroke` where its expecting a `KB KeyStroke`. This makes sense to me, but I cannot for the life of me think of a straightforward way around this. This is my first real outing with newtype's, so I was hoping there's a simple solution that I'm simply not aware of. In my understanding, the output of `cleanMask` lives in the X monad, and since I'm now in the KB monad, I have no way of getting the value out.

Would that mean the type of `waitKey` has to be `KB X KeyStroke` ?